### PR TITLE
install: autodetect cniBinDir on GKE

### PR DESCRIPTION
**Pull request by howardjohn**
Makes users not need
https://istio.io/latest/docs/setup/additional-setup/cni/#hosted-kubernetes-settings
when installing via `helm install` or `istioctl install`.

Note: `helm template` and `istioctl manifest generate` do not use
version information, so this is still needed for those cases.

This adds GKE detection. OpenShift detection is possible as well, I just
don't have a test cluster or know how to detect OpenShift

### CODE REVIEW

**Review by jacob-delgado — APPROVED** [review]

**Comment by jacob-delgado** [comment]
Should we cherry-pick this into release 1.15 and release 1.14? Possibly 1.13 as well? We'd love to get this and OCP into 1.14 ourselves.

**Comment by howardjohn** [comment]
> We'd love to get this and OCP into 1.14 ourselves.

Would be great to have the OCP equivalent added if you/someone has time!

I don't mind backport too much. Not sure how risky this is -- it feels pretty safe to me. Existing users should already have these fields explicitly set, so this change is NOP... hopefully

**Review by costinm — APPROVED** [review]
Great change !

File: operator/pkg/manifest/shared.go [hunk 1]
```diff
@@ -441,9 +441,30 @@ func getClusterSpecificValues(client kube.Client, force bool, l clog.Logger) (st
...
 	}
+	cni := getCNISettings(client)
+	if cni != "" {
+		overlays = append(overlays, cni)
+	}
 	return makeTreeFromSetList(overlays)
 }
 
+// getCNISettings gets auto-detected values based on the Kubernetes environment.
+// Note: there are other settings as well; however, these are detect inline in the helm chart.
```
Comment (kfaseela) [inline operator/pkg/manifest/shared.go:452]: nit: these are detected

File: operator/pkg/manifest/shared.go [hunk 2]
```diff
@@ -441,9 +441,30 @@ func getClusterSpecificValues(client kube.Client, force bool, l clog.Logger) (st
...
+// This ensures helm users also get them.
```
Comment (kfaseela) [inline operator/pkg/manifest/shared.go:453]: From the previous issue with fsGroup injection, this call hierarchy was basically for operator and istioctl install ? So I got a little confused with this comment.

Comment (howardjohn) [reply operator/pkg/manifest/shared.go:449]: Helm users MUST explicitly set namespace; in istioctl they do not need to. So this just detects namespace.

Comment (kfaseela) [reply operator/pkg/manifest/shared.go:449]: Okay, thanks for the clarification.

**Pull request merged**